### DISCUSSION TITLE
Question about gestures and touch.c

### DISCUSSION THREAD

**Discussion by droc12345** [labwc/labwc#2164] [Q&A]
It looks like the plumbing is there for gestures (swipe, pinch, etc), but the binding/setting of actions off of them isn't there.
Am I correct in that assumption?

And there is a touch.c in src/input, is that for touchpad or touchscreen?

**Comment by johanmalm** [comment]
@droc12345 Sorry - can't remember. Suggest looking at the sway code in first instance.

**Comment by tokyo4j** [comment]
> It looks like the plumbing is there for gestures (swipe, pinch, etc), but the binding/setting of actions off of them isn't there.
Am I correct in that assumption?

Correct.

I wrote a [daemon](https://github.com/tokyo4j/rydeen) (don't look at its code, it's a mess!) that converts libinput gesture/key events to uinput key/mouse events last year and it's working well. For example, 3-finger-down swipe sends Ctrl-W, 4-finger-down swipe sends Alt-F4, 3-finger-left/right swipe sends back/forward mouse buttons). It's not of a quality for public use though.

It's technically possible to bind those gestures to actions in labwc side, but then we'll need to implement some heuristics to determine the direction of swipe events or the threshold of swipe/pinch distance to fire actions.

> And there is a touch.c in src/input, is that for touchpad or touchscreen?

AFAIK, "touch" usually means touchscreen and touchpad is treated as a pointer device in libinput.

**Comment by concussious** [comment]
Hi, is this still the case? Labwc works great on FreeBSD, if it just had 3 finger swipe gestures it'd be perfect!

**Reply by johanmalm** [reply]
Yes. Same.

Could you confirm that you are concerned with touchscreens rather than laptop-trackpads

**Reply by concussious** [reply]
Oh sorry, no I'm concerned with laptop trackpads. There are two central gestures, three finger swipe left or right to go to the next workspace and three finger swipe up (or down, depending on how you look at it) to get to an overview. This have become a common navigation element in macos, gnome, kde, and windows (allegedly, i don't have Windows to test but this is what my friend said when I showed him), and it's really comfy.

I never thought to try it on the touchscreen, but I just did on Fedora/gnome and they work exactly the same way as on the trackpad out of the box.

**Reply by johanmalm** [reply]
See `touchpad` rather than `touch` in the documentation https://labwc.github.io/labwc-config.5.html#entry_libinput_device

There is `threeFingerDrag`: https://labwc.github.io/labwc-config.5.html#entry_libinput_device_threefingerdrag

Some clients (e.g. firefox) also handle touchpad gestures directly.

**Reply by concussious** [reply]
Can we do something in rc.xml like:
```
<touchpadbind button="threeFingerDrag" action="left">
      <action name="GoToWorkspace" />
</touchpadbind>
```

**Reply by johanmalm** [reply]
No. Suggest looking at man page referenced above.

**Reply by concussious** [reply]
I did, but I'm having trouble understanding. threeFingerDrag creates a button down event, what is the button called? How do we bind actions to that?
